synapse/third_party_rules/access_rules.py

```diff
@@ -446,8 +449,136 @@ def _on_membership_or_invite(
+            # frozen. Bail out.
+            return
+
+        user_id = event.get("sender")
+        if not user_id:
+            return
+
+        # Get every admin user defined in the room's state
+        admin_users = {
+            user
+            for user, power_level in power_level_content["users"].items()
+            if power_level >= 100
+        }
+
+        if user_id not in admin_users:
+            # This user is not an admin, ignore them
+            return
+
+        # Filter these users to only those who are actually joined or invited to the room
+        joined_members = [
+            user_id
+            for (event_type, user_id), event in state_events.items()
+            if event_type == EventTypes.Member
+            and event.membership in [Membership.JOIN, Membership.INVITE]
+        ]
+        admin_users = {user for user in admin_users if user in joined_members}
+
+        if len(admin_users) > 1:
+            # There's another admin user in, or invited to, the room
+            return
+
+        logger.info("Freezing room '%s'", event.room_id)
+
+        # Modify the existing power levels to raise all required types to 100
+        #
+        # This changes a power level state event's content from something like:
+        # {
+        #     "redact": 50,
+        #     "state_default": 50,
+        #     "ban": 50,
+        #     "notifications": {
+        #         "room": 50
+        #     },
+        #     "events": {
+        #         "m.room.avatar": 50,
+        #         "m.room.encryption": 50,
+        #         "m.room.canonical_alias": 50,
+        #         "m.room.name": 50,
+        #         "im.vector.modular.widgets": 50,
+        #         "m.room.topic": 50,
+        #         "m.room.tombstone": 50,
+        #         "m.room.history_visibility": 100,
+        #         "m.room.power_levels": 100
+        #     },
+        #     "users_default": 0,
+        #     "events_default": 0,
+        #     "users": {
+        #         "@admin:example.com": 100,
+        #     },
+        #     "kick": 50,
+        #     "invite": 0
+        # }
+        #
+        # to
+        #
+        # {
+        #     "redact": 100,
+        #     "state_default": 100,
+        #     "ban": 100,
+        #     "notifications": {
+        #         "room": 50
+        #     },
+        #     "events": {}
+        #     "users_default": 0,
+        #     "events_default": 100,
+        #     "users": {
+        #         "@admin:example.com": 100,
+        #     },
+        #     "kick": 100,
+        #     "invite": 100
+        # }
+        for key, value in power_level_content.items():
+            # Do not change "users_default", as that key specifies the default power
+            # level of new users
+            if isinstance(value, int) and key != "users_default":
+                power_level_content[key] = 100
+        power_level_content["events"] = {}
```

Comment:
rather than copying at line 474, why not build a new content object?

```suggestion
        new_content = {}
        for key, value in power_level_content.items():
            # Do not change "users_default", as that key specifies the default power
            # level of new users
            if isinstance(value, int) and key != "users_default":
                value = 100
            new_content[key] = value
        new_content["events"] = {}
```

Reply:
Also, how about PL events were the relevant keys are missing?

Reply:
True, if an event came in that had `user_default` set to `100` then the room wouldn't be very frozen.

The keys we really care about here are `state_default`, `events_default`, `users_default`. Perhaps we should in addition set the first two to `100` explicitly in the new dict?

Reply:
Also sounds like a good idea for another test case!

Reply:
I refactored the existing test in https://github.com/matrix-org/synapse-dinsic/pull/59/commits/a0daf1da647527555d8b18fb358e3bfb6d1d9b35 to create a function that takes a room ID and a dict representing the content of a power levels event, and tries to freeze a room that has that power levels contents.

I then test both the existing case (where the room has the default power levels), and and the case where a power levels event lacking `state_default` and `events_default` keys works (and asserting that those keys now exist and are set to 100).